diff --git a/autoload/leaderf/python/leaderf/gtagsExpl.py b/autoload/leaderf/python/leaderf/gtagsExpl.py
index 0a4b26c..2c34f72 100644
--- a/autoload/leaderf/python/leaderf/gtagsExpl.py
+++ b/autoload/leaderf/python/leaderf/gtagsExpl.py
@@ -483,6 +483,13 @@ class GtagsExplorer(Explorer):
         finally:
             lfCmd("echohl NONE")
 
+    def _cscope_add_db(self, dbpath, dbname='GTAGS'):
+        """auto run `cs add cs_db` if cscope_connection() is 0"""
+        if int(lfEval('cscope_connection()')) == 0:
+            if os.path.isfile(db := os.path.join(dbpath, dbname)):
+                lfCmd('set csprg=gtags-cscope')
+                lfCmd('cs add %s' % db)
+
     def _update(self, filename, single_update, auto):
         if filename == "":
             return
@@ -515,6 +522,7 @@ class GtagsExplorer(Explorer):
         elif self._isVersionControl(filename):
             if not exists:
                 self._executeCmd(root, dbpath)
+        self._cscope_add_db(dbpath)
 
     def _updateLibGtags(self, root, dbpath):
         if not self._gtagslibpath:
diff --git a/autoload/leaderf/python/leaderf/manager.py b/autoload/leaderf/python/leaderf/manager.py
index 54b19e3..697e0d0 100644
--- a/autoload/leaderf/python/leaderf/manager.py
+++ b/autoload/leaderf/python/leaderf/manager.py
@@ -136,6 +136,7 @@ class Manager(object):
         self._preview_config = {}
         self.is_autocmd = False
         self.is_ctrl_c = False
+        self._split_winid_dict = {}
         self._circular_scroll = lfEval("get(g:, 'Lf_EnableCircularScroll', 0)") == '1'
         if lfEval("has('patch-8.1.1615') || has('nvim-0.5.0')") == '0':
             lfCmd("let g:Lf_PreviewInPopup = 0")
@@ -768,7 +769,7 @@ class Manager(object):
             if show_borders:
                 config["border"] = borderchars
                 if lfEval("has('nvim-0.9.0')") == '1':
-                    config["title"] = " Preview "
+                    config["title"] = f" {title} "
                     config["title_pos"] = "center"
 
             self._createPreviewWindow(config, source, line_num, jump_cmd)
@@ -838,7 +839,7 @@ class Manager(object):
                 line = maxheight + 1
 
             options = {
-                    "title":           " Preview ",
+                    "title":           f" {title} ",
                     "maxwidth":        maxwidth,
                     "minwidth":        maxwidth,
                     "maxheight":       maxheight,
@@ -1031,8 +1032,12 @@ class Manager(object):
         """
         self._is_previewed = True
         line_num = int(line_num)
+        if isinstance(title, int):
+            title = lfEval('bufname(%d)' % title)
+        title = lfEval('fnamemodify("%s", ":~:.")' % (title or 'Preview'))
 
         if self.isPreviewWindowOpen():
+            lfCmd("call popup_setoptions(%d, %s)" % (self._preview_winid, str({'title': f' {title} '})))
             self._useExistingWindow(title, source, line_num, jump_cmd)
             return False
 
@@ -1131,6 +1136,8 @@ class Manager(object):
                     "noautocmd": 1
                     }
 
+            if lfEval("has('nvim-0.9.0')") == '1':
+                config.update({"title": f" {title} ", "title_pos": "center"})
             self._updateOptions(preview_pos, show_borders, config)
             self._createPreviewWindow(config, source, line_num, jump_cmd)
         else:
@@ -1198,7 +1205,7 @@ class Manager(object):
                 col = self._getInstance().window.col + self._getInstance().window.width - maxwidth + 1
 
             options = {
-                    "title":           " Preview ",
+                    "title":           f" {title} ",
                     "maxwidth":        maxwidth,
                     "minwidth":        maxwidth,
                     "maxheight":       maxheight,
@@ -1249,9 +1256,6 @@ class Manager(object):
                 options["border"] = borderchars
                 options["height"] -= 2
                 options["width"] -= 2
-                if lfEval("has('nvim-0.9.0')") == '1':
-                    options["title"] = " Preview "
-                    options["title_pos"] = "center"
         else:
             if preview_pos.lower() == 'cursor':
                 options["maxwidth"] = self._getInstance().window.width
@@ -2333,6 +2337,7 @@ class Manager(object):
             self._clearHighlightsPos()
             self._help_length = 0
             self._show_help = False
+            self._split_winid_dict = {}
 
     @modifiableController
     def toggleHelp(self):
@@ -2356,11 +2361,17 @@ class Manager(object):
 
             if self._getExplorer().getStlCategory() != "Help":
                 if mode == '':
-                    pass
-                elif mode == 'h':
-                    lfCmd("split")
-                elif mode == 'v':
-                    lfCmd("bel vsplit")
+                    if winid_list := list(self._split_winid_dict.values()):
+                        lfCmd("noautocmd call win_gotoid({})".format(winid_list[0]))
+                elif mode in ('h', 'v'):
+                    if winid := self._split_winid_dict.get(mode, None):
+                        lfCmd("noautocmd call win_gotoid({})".format(winid))
+                    else:
+                        if mode == 'h':
+                            lfCmd('split')
+                        else:
+                            lfCmd('leftabove vsplit')
+                        self._split_winid_dict.update({mode: int(lfEval("win_getid()"))})
 
             kwargs["mode"] = mode
             tabpage_count = len(vim.tabpages)
@@ -2409,7 +2420,7 @@ class Manager(object):
             files = []
             for i in sorted(self._selections.keys()):
                 files.append(self._getInstance().buffer[i-1])
-            if "--stayOpen" in self._arguments:
+            if ("--stayOpen" in self._arguments) and mode:
                 if self._getInstance().window.valid:
                     self._getInstance().cursorRow = self._getInstance().window.cursor[0]
                 self._getInstance().helpLength = self._help_length
@@ -2448,7 +2459,7 @@ class Manager(object):
             line_num = self._getInstance().window.cursor[0]
             need_exit = self._needExit(file, self._arguments)
             if need_exit:
-                if "--stayOpen" in self._arguments:
+                if ("--stayOpen" in self._arguments) and mode:
                     if self._getInstance().window.valid:
                         self._getInstance().cursorRow = self._getInstance().window.cursor[0]
                     self._getInstance().helpLength = self._help_length
diff --git a/autoload/leaderf/python/leaderf/rgExpl.py b/autoload/leaderf/python/leaderf/rgExpl.py
index c57b644..59b5cfe 100644
--- a/autoload/leaderf/python/leaderf/rgExpl.py
+++ b/autoload/leaderf/python/leaderf/rgExpl.py
@@ -1277,7 +1277,7 @@ class RgExplManager(Manager):
             else:
                 source = file
 
-        self._createPopupPreview("", source, line_num)
+        self._createPopupPreview(source, source, line_num)
 
     def outputToQflist(self, *args, **kwargs):
         items = self._getFormatedContents()
