diff --git a/plugin/dirdiff.vim b/plugin/dirdiff.vim
index 2389f7a..2d05c95 100644
--- a/plugin/dirdiff.vim
+++ b/plugin/dirdiff.vim
@@ -197,7 +197,8 @@ function! <SID>DirDiff(srcA, srcB)
     let DirDiffAbsSrcA = substitute(DirDiffAbsSrcA, '\\$\|/$', '', '')
     let DirDiffAbsSrcB = substitute(DirDiffAbsSrcB, '\\$\|/$', '', '')
 
-    let s:FilenameDiffWindow = tempname()
+    " let s:FilenameDiffWindow = tempname()
+    let s:FilenameDiffWindow = "/tmp/vim-dirdiff-win"
     " We first write to that file
     " Constructs the command line
     let langStr = ""
@@ -234,6 +235,7 @@ function! <SID>DirDiff(srcA, srcB)
         return
     endif
     silent exe "edit ".s:FilenameDiffWindow
+    setlocal modifiable
     echo "Defining [A] and [B] ... "
     " We then do a substitution on the directory path
     " We need to do substitution of the the LONGER string first, otherwise
@@ -414,21 +416,31 @@ function <SID>ToggleHex()
 endfunction
 
 function! <SID>DirDiffHexmode()
-    wincmd k
-    call <SID>ToggleHex()
-    wincmd l
-    call <SID>ToggleHex()
+    " Support hex mode only for existing files (not directories).
+    if filereadable(s:FilenameA)
+        call <SID>GotoFileAWindow()
+        call <SID>ToggleHex()
+    endif
+    if filereadable(s:FilenameB)
+        call <SID>GotoFileBWindow()
+        call <SID>ToggleHex()
+    endif
     " Go back to the diff window
-    wincmd j
+    call <SID>GotoDiffWindow()
 endfunction
 
 function! <SID>DirDiffWrapmode()
-    wincmd k
-    setlocal wrap!
-    wincmd l
-    setlocal wrap!
+    " Support wrap mode only for existing files (not directories).
+    if filereadable(s:FilenameA)
+        call <SID>GotoFileAWindow()
+        setlocal wrap!
+    endif
+    if filereadable(s:FilenameB)
+        call <SID>GotoFileBWindow()
+        setlocal wrap!
+    endif
     " Go back to the diff window
-    wincmd j
+    call <SID>GotoDiffWindow()
 endfunction
 
 function! <SID>EscapeFileName(path)
@@ -455,10 +467,30 @@ function! <SID>Drop(fname)
     if winid > 0
         call win_gotoid(winid)
     else
-        exe 'edit ' a:fname
+        " If the Netrw plugin is configured in tree style listing mode
+        " (i.e. g:netrw_liststyle = 3), then a buffer named 'NetrwTreeListing'
+        " is shown (i.e. active window). This happens only after the original
+        " buffer, which has the directory as its name, is focused for the first
+        " time. We work around this by comparing the 'netrw_curdir' variable of
+        " the 'NetrwTreeListing' buffer against the directory name.
+        let netrw_buffer = bufnr("NetrwTreeListing")
+        let netrw_winid = bufwinid(netrw_buffer)
+        if netrw_winid > 0 && a:fname == getbufvar(netrw_buffer, "netrw_curdir")
+            call win_gotoid(netrw_winid)
+        else
+            exe 'edit ' a:fname
+        endif
     endif
 endfunction
 
+function! <SID>GotoFileAWindow()
+    call <SID>Drop(s:FilenameA)
+endfunction
+
+function! <SID>GotoFileBWindow()
+    call <SID>Drop(s:FilenameB)
+endfunction
+
 function! <SID>GotoDiffWindow()
     call <SID>Drop(s:FilenameDiffWindow)
 endfunction
@@ -503,7 +535,7 @@ function! <SID>DirDiffOpen()
 
         if exists("s:LastMode")
             if s:LastMode == 2
-                silent exec "bd ".bufnr(previousFileA)
+                silent! exec "bd ".bufnr(previousFileA)
 
                 call <SID>Drop(previousFileB)
                 silent exec "edit ".fileToOpen
@@ -511,7 +543,13 @@ function! <SID>DirDiffOpen()
                 let previousFile = (s:LastMode == "A") ? previousFileA : previousFileB
                 call <SID>Drop(previousFile)
                 silent exec "edit ".fileToOpen
-                silent exec "bd ".bufnr(previousFile)
+                " The Netrw plugin creates an 'unlisted' buffer for directories.
+                " Running 'bd {buffer}' on an unlisted buffer produces the following error:
+                " E516: No buffers were deleted.
+                " Suppress the error with 'silent!'.
+                if fileToOpen != previousFile && filereadable(previousFile)
+                    silent! exec "bd ".bufnr(previousFile)
+                endif
             endif
         else
             silent exec "split ".fileToOpen
@@ -531,17 +569,27 @@ function! <SID>DirDiffOpen()
                 call <SID>Drop(previousFileA)
                 silent exec "edit ".s:FilenameA
                 diffthis
-                silent exec "bd ".bufnr(previousFileA)
+                if previousFileA != s:FilenameA && filereadable(previousFile)
+                    silent exec "bd ".bufnr(previousFileA)
+                endif
 
                 call <SID>Drop(previousFileB)
                 silent exec "edit ".s:FilenameB
                 diffthis
-                silent exec "bd ".bufnr(previousFileB)
+                if previousFileB != s:FilenameB && filereadable(previousFile)
+                    silent exec "bd ".bufnr(previousFileB)
+                endif
             else
                 let previousFile = (s:LastMode == "A") ? previousFileA : previousFileB
                 call <SID>Drop(previousFile)
                 silent exec "edit ".s:FilenameB
-                silent exec "bd ".bufnr(previousFile)
+                " The Netrw plugin creates an 'unlisted' buffer for directories.
+                " Running 'bd {buffer}' on an unlisted buffer produces the following error:
+                " E516: No buffers were deleted.
+                " Suppress the error with 'silent!'.
+                if previousFile != s:FilenameB && filereadable(previousFile)
+                    silent! exec "bd ".bufnr(previousFile)
+                endif
                 diffthis
 
                 " To ensure that A is on the left and B on the right, splitright must be off
@@ -603,7 +651,7 @@ function! <SID>HighlightLine()
     " This is necessary since the modified file would make the syntax
     " disappear.
     call <SID>SetupSyntax()
-    redraw
+    redraw!
 endfunction
 
 " Removes the highlight from the selected line in the diff window
@@ -620,7 +668,7 @@ function! <SID>DeHighlightLine()
     setlocal nomodifiable
     setlocal nomodified
     exe (savedLine)
-    redraw
+    redraw!
 endfunction
 
 " Returns the directory for buffer "A" or "B".  You need to be in the diff
