diff --git a/plugin/buftabline.vim b/plugin/buftabline.vim
index fccfd6d..8c1e6c3 100644
--- a/plugin/buftabline.vim
+++ b/plugin/buftabline.vim
@@ -55,6 +55,23 @@ function s:SID()
 	return matchstr(expand('<sfile>'), '<SNR>\d\+_')
 endfunction
 
+let s:mru_list = []
+function! buftabline#updatebuf()
+	let l:bufnums = buftabline#user_buffers()
+	let l:cur_buf = bufnr('%')
+	if len(l:bufnums) > 0 && index(l:bufnums, l:cur_buf) != -1
+		if len(s:mru_list) == 0
+			let s:mru_list = l:bufnums
+		endif
+
+		let s:mru_list = add(filter(s:mru_list, 'v:val != l:cur_buf'), l:cur_buf)
+		if len(s:mru_list) > g:buftabline_plug_max
+			let s:mru_list = s:mru_list[-g:buftabline_plug_max:]
+		endif
+	endif
+	return filter(l:bufnums, 'index(s:mru_list, v:val) != -1')
+endfunction
+
 let s:dirsep = fnamemodify(getcwd(),':p')[-1:]
 let s:centerbuf = winbufnr(0)
 let s:tablineat = has('tablineat')
@@ -65,7 +82,7 @@ function! buftabline#render()
 	let show_mod = g:buftabline_indicators
 	let lpad     = g:buftabline_separators ? nr2char(0x23B8) : ' '
 
-	let bufnums = buftabline#user_buffers()
+	let bufnums = buftabline#updatebuf()
 	let centerbuf = s:centerbuf " prevent tabline jumping around when non-user buffer current (e.g. help)
 
 	" pick up data on all the buffers
@@ -83,7 +100,7 @@ function! buftabline#render()
 		if strlen(bufpath)
 			let tab.path = fnamemodify(bufpath, ':p:~:.')
 			let tab.sep = strridx(tab.path, s:dirsep, strlen(tab.path) - 2) " keep trailing dirsep
-			let tab.label = tab.path[tab.sep + 1:]
+			let tab.label = currentbuf == bufnum ? tab.path : tab.path[tab.sep + 1:]
 			let pre = screen_num
 			if getbufvar(bufnum, '&mod')
 				let tab.hilite = 'Modified' . tab.hilite
@@ -96,7 +113,7 @@ function! buftabline#render()
 			let tab.label = ( show_mod ? '!' . screen_num : screen_num ? screen_num . ' !' : '!' )
 		else " unnamed file
 			let tab.label = ( show_mod && getbufvar(bufnum, '&mod') ? '+' : '' )
-			\             . ( screen_num ? screen_num : '*' )
+			\             . ( screen_num ? screen_num : '*' ) . ' --NO NAME--' . localtime()
 		endif
 		let tabs += [tab]
 	endfor
@@ -194,11 +211,23 @@ autocmd FileType qf call buftabline#update(0)
 autocmd BufDelete * call buftabline#update(str2nr(expand('<abuf>')))
 augroup END
 
-for s:n in range(1, g:buftabline_plug_max) + ( g:buftabline_plug_max > 0 ? [-1] : [] )
-	let s:b = s:n == -1 ? -1 : s:n - 1
-	execute printf("noremap <silent> <Plug>BufTabLine.Go(%d) :<C-U>exe 'b'.get(buftabline#user_buffers(),%d,'')<cr>", s:n, s:b)
-endfor
-unlet! s:n s:b
+command! -bang -nargs=* BufTablineGoto   call buftabline#action(0, '<bang>'=='!', <f-args>)
+command! -bang -nargs=* BufTablineRemove call buftabline#action(1, '<bang>'=='!', <f-args>)
+" action: 0 jump, >0 bd delete buffer
+function! buftabline#action(action, bang, ...)
+	if a:0 == 0
+		return
+	else
+		let l:bufnums = map(copy(a:000),'v:val =~ ''\d\+'' ? bufnr(v:val + 0) : bufnr(v:val)')
+	endif
+
+	let l:valid_bufnums = filter(buftabline#user_buffers(), 'index(s:mru_list, v:val) != -1')
+	for l:bufnum in filter(l:bufnums, 'v:val > 0 && v:val <= len(l:valid_bufnums)')
+		let l:cmd = 'silent! ' . (a:action > 0 ? 'bd' : 'b') . (a:bang ? '! ' : ' ') . l:valid_bufnums[l:bufnum-1]
+		exec l:cmd
+	endfor
+endfunction
+
 
 if v:version < 703
 	function s:transpile()
